File: src/main/java/com/example/rag/HelloRagChat.java
package com.example.rag;

import dev.langchain4j.data.message.AiMessage;
import dev.langchain4j.data.message.ChatMessage;
import dev.langchain4j.data.message.SystemMessage;
import dev.langchain4j.data.message.UserMessage;
import dev.langchain4j.data.segment.TextSegment;
import dev.langchain4j.model.anthropic.AnthropicChatModel;
import dev.langchain4j.model.chat.ChatModel;
import dev.langchain4j.model.embedding.EmbeddingModel;
import dev.langchain4j.model.embedding.onnx.bgesmallzhv15q.BgeSmallZhV15QuantizedEmbeddingModel;
import dev.langchain4j.model.output.Response;
import dev.langchain4j.model.chat.response.ChatResponse;
import dev.langchain4j.store.embedding.EmbeddingMatch;
import dev.langchain4j.store.embedding.EmbeddingStore;
import dev.langchain4j.store.embedding.inmemory.InMemoryEmbeddingStore;

import java.io.IOException;
import java.io.InputStream;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;
import java.util.Scanner;

/**
 * Interactive chat client for the RAG pipeline.
 */
public class HelloRagChat {

    public static void main(String[] args) {
        // 0. Cargar configuracion
        Properties props = loadProperties();
        int chunkSize = Integer.parseInt(props.getProperty("rag.chunk.size", "300"));
        int chunkOverlap = Integer.parseInt(props.getProperty("rag.chunk.overlap", "30"));
        int k = Integer.parseInt(props.getProperty("rag.retrieval.k", "3"));
        double minScore = Double.parseDouble(props.getProperty("rag.retrieval.min_score", "0.5"));
        String chatModelName = props.getProperty("rag.model.chat", "claude-haiku-4-5-20251001");

        // 1. Inicializar modelos y almacenamiento
        EmbeddingModel embeddingModel = new BgeSmallZhV15QuantizedEmbeddingModel();
        EmbeddingStore<TextSegment> store = new InMemoryEmbeddingStore<>();
        ChatModel chatModel = AnthropicChatModel.builder()
                .apiKey(requireEnv("ANTHROPIC_API_KEY"))
                .modelName(chatModelName)
                .temperature(0.0)
                .build();

        // 2. Ingesta e indexación de documentos
        IngestionService ingestionService = new IngestionService(embeddingModel, store);
        RetrievalService retrievalService = new RetrievalService(embeddingModel, store);

        System.out.println("Cargando y procesando corpus...");
        ingestionService.indexDocuments(Path.of("corpus"), chunkSize, chunkOverlap);
        System.out.println("¡Sistema RAG listo! Escribe 'salir' o 'exit' para terminar.\n");

        // 3. Inicializar historial y prompt de sistema
        List<ChatMessage> chatHistory = new ArrayList<>();
        String systemPrompt = loadSystemPrompt();
        String systemInstruction = "Eres un asistente virtual de atención al cliente de 'Casa Tortuga'.\n\n" + systemPrompt;

        // Bucle de conversacion
        Scanner scanner = new Scanner(System.in);
        while (true) {
            System.out.print("Tú: ");
            if (!scanner.hasNextLine()) {
                break;
            }
            String userInput = scanner.nextLine().trim();
            if (userInput.isEmpty()) {
                continue;
            }
            if (userInput.equalsIgnoreCase("salir") || userInput.equalsIgnoreCase("exit") || userInput.equalsIgnoreCase("quit")) {
                System.out.println("Asistente: ¡Hasta luego!");
                break;
            }

            try {
                // A. Condensar pregunta si hay historial
                String searchQuery = condenseQuery(chatHistory, userInput, chatModel);
                if (!searchQuery.equals(userInput)) {
                    System.out.printf("%n[Búsqueda RAG optimizada: '%s']%n", searchQuery);
                }

                // B. Recuperación de fragmentos
                List<EmbeddingMatch<TextSegment>> matches = retrievalService.retrieve(searchQuery, k, minScore);

                // C. Construir contexto
                String contextStr = buildContextStr(matches);

                // D. Construir lista de mensajes para el LLM
                List<ChatMessage> messages = new ArrayList<>();
                messages.add(new SystemMessage(systemInstruction));
                messages.addAll(chatHistory);

                String currentPrompt = "<context>\n" + contextStr + "</context>\n\n" +
                                       "Pregunta del usuario: " + userInput;
                messages.add(new UserMessage(currentPrompt));

                // E. Generación de respuesta
                System.out.print("Asistente: Pensando...\r");
                ChatResponse response = chatModel.chat(messages);
                String answer = response.aiMessage().text();
                System.out.println("Asistente: " + answer + "\n");

                // F. Actualizar historial de chat
                chatHistory.add(new UserMessage(userInput));
                chatHistory.add(new AiMessage(answer));

            } catch (Exception e) {
                System.out.println("\nError: " + e.getMessage() + "\n");
            }
        }
        scanner.close();
    }

    private static String condenseQuery(List<ChatMessage> chatHistory, String latestQuestion, ChatModel chatModel) {
        if (chatHistory.isEmpty()) {
            return latestQuestion;
        }

        StringBuilder historyBuilder = new StringBuilder();
        for (ChatMessage msg : chatHistory) {
            String role = (msg instanceof UserMessage) ? "Usuario" : "Asistente";
            String text = "";
            if (msg instanceof UserMessage userMsg) {
                text = userMsg.singleText();
            } else if (msg instanceof AiMessage aiMsg) {
                text = aiMsg.text();
            }
            if (text.contains("Pregunta del usuario:")) {
                text = text.substring(text.lastIndexOf("Pregunta del usuario:") + "Pregunta del usuario:".length()).trim();
            }
            historyBuilder.append(role).append(": ").append(text).append("\n");
        }

        String condensePrompt =
            "Dado el siguiente historial de conversación y una nueva pregunta del usuario, " +
            "crea una consulta de búsqueda independiente y optimizada en español para buscar en una base de datos vectorial.\n" +
            "La consulta debe ser concisa y resolver pronombres o referencias contextuales.\n\n" +
            "Historial de conversación:\n" + historyBuilder.toString() + "\n" +
            "Nueva pregunta: " + latestQuestion + "\n\n" +
            "Consulta de búsqueda optimizada (responde SOLO con la consulta):";

        String condensed = chatModel.chat(condensePrompt);
        if (condensed != null) {
            condensed = condensed.trim().replaceAll("^\"|\"$|^'|'$", "");
        }
        return condensed;
    }

    private static String buildContextStr(List<EmbeddingMatch<TextSegment>> matches) {
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < matches.size(); i++) {
            TextSegment segment = matches.get(i).embedded();
            String source = segment.metadata().getString("file_name");
            sb.append("<doc id=\"").append(i + 1).append("\"")
                    .append(" source=\"").append(source != null ? source : "desconocido").append("\">\n")
                    .append(segment.text()).append("\n")
                    .append("</doc>\n");
        }
        return sb.toString();
    }

    private static Properties loadProperties() {
        Properties props = new Properties();
        try (InputStream input = HelloRagChat.class.getClassLoader().getResourceAsStream("application.properties")) {
            if (input == null) {
                System.out.println("(!) No se encontro application.properties. Usando valores por defecto.");
                return props;
            }
            props.load(input);
        } catch (IOException ex) {
            System.err.println("(!) Error cargando application.properties: " + ex.getMessage());
        }
        return props;
    }

    private static String loadSystemPrompt() {
        Path path = Path.of("system_prompt.txt");
        if (!Files.exists(path)) {
            path = Path.of("../system_prompt.txt");
        }
        if (!Files.exists(path)) {
            path = Path.of("../../system_prompt.txt");
        }
        try {
            return Files.readString(path).strip();
        } catch (IOException e) {
            throw new RuntimeException("No se pudo leer el archivo system_prompt.txt en " + path.toAbsolutePath(), e);
        }
    }

    private static String requireEnv(String name) {
        String value = System.getenv(name);
        if (value == null || value.isBlank()) {
            throw new IllegalStateException("Variable de entorno '" + name + "' no definida.");
        }
        return value;
    }
}

File: python/hello_rag_chat.py
import os
import glob
from langchain_core.documents import Document
from langchain_text_splitters import RecursiveCharacterTextSplitter
from langchain_huggingface import HuggingFaceEmbeddings
import chromadb
from langchain_chroma import Chroma
from langchain_anthropic import ChatAnthropic
from langchain_core.messages import SystemMessage, HumanMessage, AIMessage

# 0. Configuración básica
CHUNK_SIZE = 300
CHUNK_OVERLAP = 30
K = 3
MIN_SCORE = 0.5
CHAT_MODEL_NAME = "claude-haiku-4-5-20251001"

def require_env(name: str) -> str:
    val = os.getenv(name)
    if not val or not val.strip():
        raise ValueError(f"Variable de entorno '{name}' no definida.")
    return val

def load_documents(corpus_dir: str) -> list[Document]:
    documents = []
    for filepath in glob.glob(os.path.join(corpus_dir, "*.md")):
        with open(filepath, "r", encoding="utf-8") as f:
            content = f.read()
        doc = Document(
            page_content=content,
            metadata={"file_name": os.path.basename(filepath)}
        )
        documents.append(doc)
    return documents

def condense_query(chat_history, latest_question, chat_model) -> str:
    """
    Toma el historial del chat y la última pregunta del usuario, y genera
    una consulta de búsqueda independiente que contiene todo el contexto.
    """
    if not chat_history:
        return latest_question

    history_parts = []
    for msg in chat_history:
        role = "Usuario" if isinstance(msg, HumanMessage) else "Asistente"
        # Limpiamos posibles metadatos del prompt del usuario
        content = msg.content
        if "Pregunta del usuario:" in content:
            content = content.split("Pregunta del usuario:")[-1].strip()
        history_parts.append(f"{role}: {content}")
    history_str = "\n".join(history_parts)

    condense_prompt = (
        "Dado el siguiente historial de conversación y una nueva pregunta del usuario, "
        "crea una consulta de búsqueda independiente y optimizada en español para buscar en una base de datos vectorial.\n"
        "La consulta debe ser concisa y resolver pronombres o referencias contextuales.\n\n"
        f"Historial de conversación:\n{history_str}\n\n"
        f"Nueva pregunta: {latest_question}\n\n"
        "Consulta de búsqueda optimizada (responde SOLO con la consulta):"
    )
    
    condensed_query = chat_model.invoke(condense_prompt).content.strip()
    # Limpiamos comillas si el modelo las añade
    condensed_query = condensed_query.strip('"').strip("'")
    return condensed_query

def build_context_str(matches) -> str:
    context_parts = []
    for idx, (doc, _) in enumerate(matches):
        source = doc.metadata.get("file_name", "desconocido")
        context_parts.append(f'<doc id="{idx + 1}" source="{source}">\n')
        context_parts.append(doc.page_content + "\n")
        context_parts.append("</doc>\n")
    return "".join(context_parts)

def main():
    api_key = require_env("ANTHROPIC_API_KEY")

    # 1. Inicializar modelos
    embeddings = HuggingFaceEmbeddings(model_name="BAAI/bge-small-en-v1.5")
    chroma_client = chromadb.HttpClient(host="localhost", port=8000)
    chroma_client.delete_collection("casa-tortuga")
    vector_store = Chroma(
        collection_name="casa-tortuga",
        embedding_function=embeddings,
        client=chroma_client,
        collection_metadata={"hnsw:space": "cosine"},
    )
    chat_model = ChatAnthropic(model=CHAT_MODEL_NAME, api_key=api_key, temperature=0)

    # 2. Ingesta e indexación de documentos
    script_dir = os.path.dirname(os.path.abspath(__file__))
    corpus_dir = os.path.join(script_dir, "..", "corpus")
    
    print("Cargando y procesando corpus...")
    raw_docs = load_documents(corpus_dir)
    splitter = RecursiveCharacterTextSplitter(chunk_size=CHUNK_SIZE, chunk_overlap=CHUNK_OVERLAP)
    chunks = splitter.split_documents(raw_docs)
    vector_store.add_documents(chunks)
    print(f"Indexados {len(chunks)} vectores en ChromaDB.")
    print("¡Sistema RAG listo! Escribe 'salir' o 'exit' para terminar.\n")

    # 3. Inicializar historial de chat y leer prompt de sistema
    chat_history = []
    system_prompt_path = os.path.join(script_dir, "..", "system_prompt.txt")
    try:
        with open(system_prompt_path, "r", encoding="utf-8") as f:
            system_instruction = f.read().strip()
    except Exception as e:
        raise FileNotFoundError(f"No se pudo leer el archivo de prompt de sistema en {system_prompt_path}: {e}")

    # Le añadimos la definición de persona al inicio
    system_instruction = f"Eres un asistente virtual de atención al cliente de 'Casa Tortuga'.\n\n{system_instruction}"

    # Bucle de conversación
    while True:
        try:
            user_input = input("Tú: ").strip()
            if not user_input:
                continue
            if user_input.lower() in ["salir", "exit", "quit"]:
                print("Asistente: ¡Hasta luego!")
                break

            # A. Condensar pregunta si hay historial
            search_query = condense_query(chat_history, user_input, chat_model)
            if search_query != user_input:
                print(f"\n[Búsqueda RAG optimizada: '{search_query}']")

            # B. Recuperación de fragmentos
            results = vector_store.similarity_search_with_relevance_scores(search_query, k=K)
            matches = [m for m in results if m[1] >= MIN_SCORE]

            # C. Construir contexto
            context_str = build_context_str(matches)

            # D. Construir lista de mensajes para el LLM
            messages = [SystemMessage(content=system_instruction)]
            # Añadimos el historial previo
            messages.extend(chat_history)
            # Añadimos el mensaje actual estructurado con el contexto RAG de este turno
            current_prompt = (
                f"<context>\n{context_str}</context>\n\n"
                f"Pregunta del usuario: {user_input}"
            )
            messages.append(HumanMessage(content=current_prompt))

            # E. Generación de respuesta
            print("Asistente: Pensando...", end="\r")
            response = chat_model.invoke(messages)
            answer = response.content
            print(f"Asistente: {answer}\n")

            # F. Actualizar historial de chat (usando la pregunta original limpia y la respuesta)
            chat_history.append(HumanMessage(content=user_input))
            chat_history.append(AIMessage(content=answer))

        except KeyboardInterrupt:
            print("\nAsistente: ¡Hasta luego!")
            break
        except Exception as e:
            print(f"\nError: {e}\n")

if __name__ == "__main__":
    main()
